docs/md/logging.md

```diff
@@ -0,0 +1,68 @@
+# WFL Logging
+
+## Usage
+We use `clojure.tools.logging`.
+
+Require it like any other dependency:
+```clojure
+(ns "..."
+  (:require
+    ...
+    [clojure.tools.logging :as log]
+    ...))
+
+(log/info "Hello!")
+```
+
+Full documentation is available [here](http://clojure.github.io/tools.logging/#clojure.tools.logging),
+but the primary functions we use are these:
+
+- `log/error` and `log/errorf` for failures and uncaught exceptions
+- `log/warn` and `log/warnf` for issues that could be recovered from
+- `log/info` and `log/infof` for recording normal interaction
+- `log/debug` and `log/debugf` for more granular output that might be 
+added during debugging
+
+The first function of each level accepts any number of arguments and 
+concatenates them with spaces, while the second variant uses format 
+strings. All functions can also accept a throwable before other normal 
+arguments to log a specific exception.
+
+
+## Testing
+clojure.tools.logging provides a [test namespace](http://clojure.github.io/tools.logging/#clojure.tools.logging.test).
+
+An example of usage is in `test/zero/unit/logging_test.clj`. 
+That file doesn't actually test our code, but rather our dependencies:
+it helps catch issues that might arise from clojure.tools.logging (or SLF4J, which
+it currently delegates to) having trouble finding a logging implementation via
+service loaders and the like. It isn't bulletproof--it is theoretically possible
+for tests to use a different classpath than other execution--but it is a good smoke
+test that in that any failures there would require remediation in our own codebase.
+
+## Under the Hood
+WFL's logging works as follows:
+
+1. clojure.tools.logging is imported and used directly
+   - Why clojure.tools.logging? It is Clojure-native so it has intuitive syntax
+   - Why not wrap or delegate to it? It is written in macros so delegation
+     is messier, the macros themselves just delegate to other logging
+     implementations anyway, and its interface is so generic that we'd be
+     repeating it very closely if we were to wrap it
```

Comment:
This is what makes wrapping such an easy task.
But not wrapping, you've injected a external dependency into almost every single file in workflow launcher. It's quite possible that we may want to change this in the future.
As a general software engineering tip, try to minimise how much of your code has to know about the dependencies - it's much easier to change code in one place than several.

Reply:
I agree that wrapping is often a useful pattern, but there's a reason we don't bother wrapping `clojure.string` or `clojure.pprint`. The key thing to understand about `clojure.tools.logging` is it is literally just a wrapper itself that abstracts over more logging frameworks than even SLF4J supports.

Even if Broad decided to develop its own JVM logging library from scratch, a trivial SLF4J "impl" (like what Log4j2 uses right now) would make it work with `clojure.tools.logging`. I'm having trouble coming up with a scenario where the Clojure wrapper built for exactly this purpose doesn't offer enough flexibility and we want to wrap it again.

Reply:
Yes indeed, you're not going to wrap `string` or any other built in data type. I was kinda thinking someone might raise this lol! I really wouldn't group logging capabilities with core language types like `string` though.

In this case, I meant switching to something like timbre. I guess this is unlikely.

Reply:
Totally fair. Timbre supports [both clojure.tools.logging and SLF4J](https://github.com/ptaoussanis/timbre#a-pure-clojurescript-logging-library), so we could switch to it super easily.